src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_DecisionDagRewriter.cs

```diff
@@ -410,53 +412,510 @@ private void GenerateTest(BoundExpression test, BoundDecisionDagNode whenTrue, B
             private bool GenerateSwitchDispatch(BoundDecisionDagNode node, HashSet<BoundDecisionDagNode> loweredNodes)
             {
                 Debug.Assert(!loweredNodes.Contains(node));
-
-                // We only generate a switch dispatch if we have two or more value tests in a row
-                if (!(node is BoundTestDecisionDagNode firstTestNode &&
-                     firstTestNode.Test is BoundDagValueTest firstTest &&
-                     firstTestNode.WhenFalse is BoundTestDecisionDagNode whenFalse &&
-                     whenFalse.Test is BoundDagValueTest secondTest &&
-                     !loweredNodes.Contains(whenFalse) &&
-                     !this._dagNodeLabels.ContainsKey(whenFalse) &&
-                     firstTest.Input.Equals(secondTest.Input) &&
-                     firstTest.Input.Type.IsValidV6SwitchGoverningType()))
-                {
-                    // https://github.com/dotnet/roslyn/issues/12509 Could optimize float, double, decimal value switches. For now use if-then-else
+                if (!CanGenerateSwitchDispatch(node, loweredNodes))
                     return false;
+
+                var input = ((BoundTestDecisionDagNode)node).Test.Input;
```

Comment:
>var input = ((BoundTestDecisionDagNode)node).Test.Input; [](start = 16, length = 56)

Should `CanGenerateSwitchDispatch()` return this value?

Reply:
I could thread it through the two levels of calls but the savings are tiny and I'd rather keep things as simple as possible at this stage.

---
In reply to: [390489766](https://github.com/dotnet/roslyn/pull/42313#discussion_r390489766) [](ancestors = 390489766)